### PR TITLE
@zephraph => [Performance] Instrument TTI Metric

### PR DESCRIPTION
**Pull request by mzikherman**
cc @eessex 

Based on https://developers.google.com/web/fundamentals/performance/user-centric-performance-metrics , this adds the described metric to volley reporting. I was curious to see the kinds of values that would be reported when playing around locally!

For artist pages, this seemed to range between 4-6 seconds locally.
For artwork pages, it ranged from 1-6 seconds.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please don't include .js files, we want to be using TypeScript found: <a href="https://github.com/mzikherman/force/blob/measure_tti/src/desktop/components/main_layout/tti-init.js">tti-init.js</a>.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 82ff88e26035ffae93944c0a550d3fb194cb5b20
</p>

**Comment by joeyAghion** [comment]
* Early on, we discussed TTI as being "expensive" to calculate on the client, and so we might need to use sampling to avoid impacting our users too much. Is that still the case?
* It looks like this is just initialized once. Would it report a metric for client-side "navigations?"
* Is this Chrome-only?

**Comment by zephraph** [comment]
There are two approaches to measuring TTI. One is naive and highly expensive the other is relatively performant but with an overall low level of support (definitely chrome only). I _think_ this polyfill falls into  that latter category.

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please don't include .js files, we want to be using TypeScript found: <a href="https://github.com/mzikherman/force/blob/measure_tti/src/desktop/components/main_layout/tti-init.js">tti-init.js</a>.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2ff0fc2db6be642e1bfbaa3e60193c7bbd7fa71e
</p>

**Comment by mzikherman** [comment]
> Early on, we discussed TTI as being "expensive" to calculate on the client, and so we might need to use sampling to avoid impacting our users too much. Is that still the case?

I'm not 100% sure, investigating the `tti-polyfill` lib reveals a bunch of sophistication, and timers. I would defer to @zephraph in the meantime, although I'm not finding mention of any worry about client-side perf. in any blog posts or documentation (so far).

> It looks like this is just initialized once. Would it report a metric for client-side "navigations?"

Good question! Not sure...I'm not sure if that's a blocker to releasing TTI instrumentation generally, although it might be a blocker for releasing client-side routing, since we'd want to be setup to measure all the things. This is worth thinking through now, but can still be a follow-up?

> Is this Chrome-only?

Also good question! It looks like the polyfill works with any modern browser that implements [`PerformanceObserver`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver) and [`PerformanceLongTaskTiming`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceLongTaskTiming).

It appears as though IE doesn't work with the former, and Firefox doesn't work with the latter, so Chrome and Safari are left?

**Comment by joeyAghion** [comment]
Agree about the client-side routing issue not necessarily being a blocker. Just wanted to be aware.

If this isn't expensive and so doesn't require sampling, are there other caveats or considerations I'm missing? My impression was that measuring TTI was _not_ going to be as straightforward as importing a polyfill library.

File: src/lib/userPerformanceMetrics.ts [hunk 1]
```diff
@@ -159,3 +161,8 @@ export function getDomComplete() {
   if (!timingAvailable) return null
   return sanitizedMetrics(perf.timing.requestStart, perf.timing.domComplete)
 }
+
+export async function getTTI() {
+  if (!timingAvailable) return null
```
Comment (mzikherman) [inline src/lib/userPerformanceMetrics.ts]: This check probably needs to be more restrictive, should check the corresponding API, or maybe browser versions.

File: src/desktop/components/main_layout/tti-init.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+function ttiInit() {
```
Comment (zephraph) [inline src/desktop/components/main_layout/tti-init.js]: You should probably could just unwrap this. The function wrapper seems a little unnecessary.

File: src/lib/userPerformanceMetrics.ts (hunk 1 above)
Comment (zephraph) [reply src/lib/userPerformanceMetrics.ts]: I don't think it really needs to be any more restrictive. The polyfill itself does all the feature checking that's necessary.

Comment (mzikherman) [reply src/lib/userPerformanceMetrics.ts]: Indeed: https://github.com/GoogleChromeLabs/tti-polyfill/blob/4b5058efb695f2f51aa6f586d9a46dc9c0207f30/src/index.js#L35

**Comment by mzikherman** [comment]
This is ready, verified that this reports in mobile chrome, and in unsupported browsers will not include the metric in the Volley payload.

**Comment by zephraph** [comment]
All right, let's ship it. :shipit:

**Pull request merged**